defrag/defrag.go

```diff
@@ -0,0 +1,276 @@
+// It use a map of all the running flows
+func (d *IPv4Defragmenter) DefragIPv4(ip *layers.IPv4) (bool, error) {
+	// check if we need to defrag
+	if st := d.DontDefrag(ip); st == true {
+		return true, nil
+	}
+	// perfom security checks
+	st, err := d.SecurityChecks(ip)
+	if err != nil || st == false {
+		return st, err
+	}
+
+	// ok, got a fragment
+	debug.Printf("defrag: got ip.Id=%d ip.FragOffset=%d ip.Flags=%d\n",
+		ip.Id, ip.FragOffset*8, ip.Flags)
+
+	// do we already has seen a flow between src/dst with that Id
+	ipf := *NewIPv4Flow(ip)
+	var fl *FragmentList
+	var exist bool
+	d.Lock()
+	fl, exist = d.IPFlows[ipf]
+	if !exist {
+		debug.Printf("defrag: creating a new flow\n")
+		fl = new(FragmentList)
+		d.IPFlows[ipf] = fl
+	}
+	d.Unlock()
+	// insert, and if final build it
+	st, err = fl.Insert(ip)
+
+	// at last, if we hit the maximum frag list len
+	// without any defrag success, we just drop everything and
+	// raise an error
+	if st == false && fl.List.Len()+1 > IPv4MaximumFragmentListLen {
+		d.Lock()
+		fl = new(FragmentList)
+		d.IPFlows[ipf] = fl
+		d.Unlock()
+		return false, fmt.Errorf("defrag: Fragment List hits its maximum"+
+			"size(%d), without sucess. Flushing the list",
+			IPv4MaximumFragmentListLen)
+	}
+
+	return st, err
+}
+
+// DiscardOlderThan forgets FragmentList without any activity since
+// time t. It returns the number of FragmentList aka number of
+// fragment packets it has discarded.
+func (d *IPv4Defragmenter) DiscardOlderThan(t time.Time) int {
+	var nb int
+	d.Lock()
+	for k, v := range d.IPFlows {
+		if v.LastSeen.Before(t) {
+			nb = nb + 1
+			delete(d.IPFlows, k)
+		}
+	}
+	d.Unlock()
+	return nb
+}
+
+// DontDefrag returns true if the IPv4 packet do not need
+// any defragmentation
+func (d *IPv4Defragmenter) DontDefrag(ip *layers.IPv4) bool {
+	// don't defrag packet with DF flag
+	if ip.Flags&layers.IPv4DontFragment != 0 {
+		return true
+	}
+	// don't defrag not fragmented ones
+	if ip.Flags == 0 && ip.Id == 0 {
+		return true
+	}
+	return false
+}
+
+// SecurityChecks performs the needed security checks
+func (d *IPv4Defragmenter) SecurityChecks(ip *layers.IPv4) (bool, error) {
+	// don't allow too big fragment offset
+	if ip.FragOffset*8 > IPv4MaximumFragmentOffset {
+		return false, fmt.Errorf("defrag: fragment offset too big "+
+			"(handcrafted? %d > %d)", ip.FragOffset*8, IPv4MaximumFragmentOffset)
+	}
+
+	// don't allow fragment that would oversize an IP packet
+	if ip.FragOffset*8+ip.Length > IPv4MaximumSize {
+		return false, fmt.Errorf("defrag: fragment will overrun "+
+			"(handcrafted? %d > %d)", ip.FragOffset*8+ip.Length, IPv4MaximumSize)
+	}
+
+	return true, nil
+}
+
+// FragmentList holds a container/list used to contains IP
+// packets/fragments.  It stores internal counters to track the
+// maximum total of byte, and the current length it has received.
+// It also stores a flag to know if he has seen the last packet.
+type FragmentList struct {
```

Comment:
Let's make sure to only export the stuff we expect other folks to use.  I specifically think we don't want to export this, because there are many possible ways of defrag'ing packets, and we might want to switch which we use in the future away from a linked-list approach.